### ISSUE TITLE
Industrial SOTA Data Pipeline

### ISSUE THREAD

**Issue by crutcher** [tracel-ai/burn#3363]
Burn's *default* dataset / data augmentation / loading pipeline should target existing SOTA pipeline features by default.

much of the `torch` development is split by putting all the industrial training stuff into `lightning`; which splits attention. It would be nice if we didn't do this.

The individual features are less important than a data pipeline and train loop which can host them by default.

### File Tree Datasets

"filetree/folder walking datasets" should be a general plugin mechanism, not a private implementation feature of `ImageFolderDataset`.

The scan machinery should be pluggable, to support things like S3; and the inside of zip files.

The way in which paths are bound to annotations should also be pluggable.

The indexing to a dataset should be an entire layer, before anything is loaded.

### Finite vs Periodic Datasets

Some datasets are finite, there are exactly X things  in them.
Some datasets are periodic; they are derived from a source, with some stochastic modification.

#### Idempotent Epochs

It would be very nice to be able to extract a sample period, representing an epoch, from a periodic dataset, with idempotent sampling machinery (reading the same indexed sample produces the exact same load); for a host of distribution reasons.

This might be done by pulling a "epoch_sample() => DataSample" interface; but it would be nice if that machinery had access to things like temperature, and the current computed metrics surface.

### Dataset Combinators and Augmentors

A variety of dataset combinators are needed; simple weighted union, fractional sampling, stochastic augmentation, etc.

### Dataloaders *after* dataset => datasample pipeline

Once we can apply the dataset graph, and extract a datasample for *this* epoch, the loader is responsible (as now) for loading it.

**Comment by crutcher** [comment]
@laggui

**Comment by antimora** [comment]
This also be useful for audio data pipeline. 

CC @nathanielsimard

**Comment by crutcher** [comment]
I've been thinking about this, and taking notes and comparing to previous systems. I'm going to sketch out a pattern here, and deliberately divorce myself from the existing names.

Consider a Apache Beam-like PCollection (a *promise* of a distributed, parallel collection).

## DataSchedule<M, V>

Let's say that we have a thing called a `DataSchedule<M, V>`.
This object has an exact size (`len()`); and a random-accessable index vector (`Vec<M>`).
`M` is a meta recipe for one instance, it is not required to be unique in the schedule; the index list *is* assumed to be in-memory.
the schedule has a mapper `M -> V`; which is assumed to be potentially expensive.

For a given schedule-run (say, one epoch), a schedule-provider gives a schedule; and that schedule is passed off to the dataloader machinery (and whatever logging/sampling is built into that).

By the time a DataSchedule is built, there are no more global choices (the number of items is fixed; etc).

## DataSource<M, V>

A data source is a thing which can *produce* a `DataSchedule<M, V>` upon demand.
It has a `.period() -> usize` and a `.get(meta: M)`.
It has a `.schedule(options: ScheduleOptions, rnd: &mut Rnd) -> DataSchedule<M, V>` method; the options include a `fraction: f32` for under and over sampling.

A DataSource can be thought of as like a `SQL` `select ? from ...`; we're not altering the table at all; just defining it.

### High-Order DataSource

A High-Order DataSource is a combinator which combines or manipulates the schedules produced by other DataSources. Examples include:
* Shuffle the schedule produced by a wrapped DataSource.
* Join multiple wrapped DataSources, and shuffle their schedules.

## DataScheduleTransform<K1, V1, K2, V2>

This is a transform *on* an entire schedule. This transform can change the number of items, their ordering, and attach additional mapping processing steps. Process-based filtering, and injection of things like `temperature` go here.

The main way in which this differs from a high-level re-use of a `DataSource` is that we want the train-loop to inject meaningful state information into this, maybe. Baring that, this could just be a DataSource.

# Summary

A user builds a tree of low and high-level DataSource objects (possibly config driven) in much the same way they would bind a SQL expression to tables.

The user then defines a DataScheduleTransform, for how, on a *given* epoch, the schedule produced naively by the final DataSource would be globally re-written into a fixed, idempotent lazy schedule (it's still meta-only, no load/aug work has been done; but the transform has attached the callbacks *for* the load/aug work to the `M -> V` hooks).

The schedule is passed off to the parallel/distributed dataloader, which is stupid, and just does the work (like now); though maybe with a bit more structure than an `.iter()`.

# Potential Future Work

## Schedule Watchers

Schedules, once constructed, are interesting metric / logging targets. And the requirement that they be idempotent helps with that a lot.

## Column-Oriented Data

All of this assumes that we shove all of our *Data* into, ultimately, a single type. Everything gets a lot simpler if there can be multiple independent columns; loaded in parallel, with their own metadata and build mappers.

**Comment by crutcher** [comment]
In case it is not clear, the *work* is done by the mapper function when the DataLoader actually pulls an item from the schedule. Any of the previous DataSources could have provided the associated Mapper, or could have re-written the Meta; and the DataScheduleTransform is assumed to run at a time when it can also make choices based upon train characteristics (LR, temp, etc); but everything before then is either adjustments to the meta, or updates to the bound mapper.

**Comment by nathanielsimard** [comment]
I think most of the things you mentionned can be done with the current abstractions. The `Dataset` trait abstracts all mappers, samplers and data sources. Most of those things are done lazily for best performance. Some features are still missing, like the "filetree/folder walking datasets", S3 and other data formats. We could have more samplers and more mappers, but with the `ComposedDataset` I'm unsure what's the limitation of the current architecture?
